SSL: support for compressed server certificates with BoringSSL.

pluknet opened the pull request:
BoringSSL provides two callbacks to register on SSL context for each compression algorithm, which is used to send and receive compressed server certificates, as appropriate.  This change adds support for compression with zlib, enabled with the ssl_certificate_compression directive.  Compressed certificates are cached in SSL context on the first callback invocation.

pluknet committed:
SSL: support for compressed server certificates with BoringSSL.

BoringSSL/AWS-LC provide two callbacks for each compression algorithm,
which may be used to compress and decompress certificates in runtime.
This change implements compression support with zlib, as enabled with
the ssl_certificate_compression directive.  Compressed certificates
are stored in certificate exdata and reused in subsequent connections.

Notably, AWS-LC saves an X509 pointer in SSL connection, which allows
to use it from SSL_get_certificate() for caching purpose.  In contrast,
BoringSSL reconstructs X509 on-the-fly, though given that it doesn't
support multiple certificates, always replacing previously configured
certificates, we use the last configured one from ssl->certs, instead.

bavshin-f5 commented on src/event/ngx_event_openssl.c:
```diff
@@ -694,6 +734,129 @@ ngx_ssl_certificate_compression(ngx_conf_t *cf, ngx_ssl_t *ssl,
+
+    if (comp != NULL) {
+        return CBB_add_bytes(out, comp->data, comp->len);
+    }
```
This assumes that the compressed chain will never change during the SSL_CTX lifetime. I've confirmed the following with BoringSSL and AWS-LC:

* a single certificate/key pair is allowed on the context
* compression will not be negotiated for certificates set in the certificate callback

The former is a documented behavior, but I can't find what guarantees the latter.

pluknet replied:
* actually, a single pair of each type (RSA, ECDSA, ...) (though at least BoringSSL doesn't support multiple certificates)
* ngx_ssl_certificate_compression() is invoked only for statically configured certificates as originally done in 251444fcf; this was done for a reason:
-- OpenSSL doesn't provide API to compress certificates in runtime in a non-ugly way
-- generally compression for "dynamic" certificates seems to bring more harm than good

bavshin-f5 replied:
> * actually, a single pair of each type (RSA, ECDSA, ...) (though at least BoringSSL doesn't support multiple certificates)

Hm. I rechecked and it appears that AWS-LC does support multiple certs while using the same compression API. So it is possible to invoke the callback on the same context with a different input data.

>     * ngx_ssl_certificate_compression() is invoked only for statically configured certificates as originally done in [251444f](https://github.com/nginx/nginx/commit/251444fcf4434bfddbe3394a568c51d4f7bd857f)

Got it, I missed the scope of the `ngx_ssl_certificate_compression` call. That's one scenario we don't have to care about.

pluknet replied:
AWS-LC was not taken into account originally back to 2023, so were not multiple certificates.
Regardless, now fixed in [4de1157](https://github.com/nginx/nginx/commit/4de115779f4d18c8794bc0a0c176cd01e363ceeb).

bavshin-f5 replied:
Looks good and beneficial even for BoringSSL: after the SSL cache changes, we can have the same X509 chain attached to multiple servers. Now the cached compressed message can be shared as well.

Consider adding `free_func` to the exdata. While not necessary for production use, it will suppress the noise from valgrind and sanitizers and eliminate a potential leak with `master_process off`.

***

Aside from the certificate chain, two more things may affect the uncompressed Certificate message: OCSP stapling and Signed Certificate Timestamp. Both extensions are added on request from the client (https://github.com/google/boringssl/blob/main/ssl/tls13_both.cc#L449).

We don't support the latter, and the [third-party module](https://github.com/grahamedgecombe/nginx-ct) seems abandoned and incompatible with our recent releases. We do support OCSP stapling with aws-lc, though.

pluknet replied:
We generally don't bother with freeing worker process memory on exit for simplicity, as it will be naturally freed by operating system. Existing examples include hcf->points in the upstream hash balancer and cycle connections.

A specific of OCSP stapling, when using OCSP responders, is in limitations in the original OpenSSL API, which doesn't allow to send a stapled response in the first connection without potentially blocking. This means that the 1st connection is normally served without a stapled response.
Notably, OCSP response is sent in the Certificate message itself (unlike in previous TLS versions), so caching such certificates unconditionally means a factually disabled stapling.
Fixed in [cd8f5d5](https://github.com/nginx/nginx/commit/cd8f5d52213d93e61d48d04c1fb246bed5241473) 

Not sure about SCT, but since we don't support it I would rather not care.  At least, BoringSSL doesn't provide callbacks to register, so it seems to be set by external means (see SSL_CTX_set_signed_cert_timestamp_list).

bavshin-f5 replied:
> We generally don't bother with freeing worker process memory on exit for simplicity, as it will be naturally freed by operating system. Existing examples include hcf->points in the upstream hash balancer and cycle connections.

We actually do for the most cases. The specific examples you mention require an ugly cleanup code that was rejected, but the list of LeakSanitizer suppressions I have to maintain is rather small.

***

Interesting fact about OpenSSL: as you're undoubtedly aware, it requires building the `CompressedCertificate` messages in advance. It also prefers sending such pre-compressed message if it's possible to negotiate compression.
As a result, `ssl_certificate_compression on` and `ssl_stapling on` cannot work together: the OCSP Status extension will never be sent to the clients that request both. Now, add OCSP Must-Staple to the cert and you'll get a subtly broken configuration.
This can be verified with a trivial change in `ssl_stapling.t` (and slightly less trivial decoding of Compressed Certificate Message from pcap, because Wireshark treats it as an opaque data).

I also see two potential issues with the latest diff:
* OCSP status from responder is refreshed regularly; there's no invalidation happening so we can be sending stale data
* [RFC6066 § 8](https://datatracker.ietf.org/doc/html/rfc6066#section-8):
  > Note in addition that a server MUST NOT send the "CertificateStatus"
  > message unless it received a "status_request" extension in the client
  > hello message and sent a "status_request" extension in the server
  > hello message.
  
  and [RFC8446 § 4.4.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2), which is more relevant for us, but carries a similar message:
  
  > Extensions in the Certificate message from the server MUST
  > correspond to ones from the ClientHello message.

  The client is not obligated to send the status request and the specifications forbid us from sending an unsolicited status from cache (I have no idea how this supposed to interact with must-staple).

tl;dr: I'm not sure if that's the best way to handle the issue, but I suggest to reject the combination of `ssl_compressed_certificate` + `ssl_stapling` at configuration time instead of increasing the complexity of the callback.

pluknet replied:
> As a result, ssl_certificate_compression on and ssl_stapling on cannot work together

That's a correct observation, made directives explicitly mutually exclusive.

Further, the previous move to cert ex_data broke caching, because X509 are dup'ed on SSL_new().
The fix seems to lookup a corresponding X509 entry in SSL_CTX.

While testing, I also found that repeatitive loading of the same certificate type makes extraneous expansion of `ssl->certs`, this needs to be fixed separately.

Moving to SSL_CTX makes again freeing ex_data relatively easy, through the cleanup handler, so I put it back.

See the 0faee9cb7 tip.

pluknet replied:
> extraneous expansion

~addressed as part of https://github.com/nginx/nginx/pull/740, no changes in this PR~

Edit: reverted, not going to work

pluknet commented:
Fixed log level, resolved conflicts.

pluknet commented:
Added support for AWS-LC by moving to certificate exdata, related cleanup and simplifications.

pluknet commented:
Fixed OCSP stapling (AWS-LC), miscellaneous changes.

pluknet committed:
SSL: disabled using certificate compression with OCSP stapling.

OCSP response in TLSv1.3 is sent in the Certificate message.  This
is incompatible with pre-compression of the configured certificates.

pluknet commented:
- disabled using certificate compression with OCSP stapling
- fixed caching

pluknet commented:
Minor cleanup (getting rid of crocodiles).

pluknet commented:
Changes: cert lookup fixes
- removed X509_get0_pubkey() complexity for BoringSSL
  to associate cert with ssl_ctx, this never worked anyway;
  cert lookup is rewritten to use the last configured cert in ssl_ctx;
  BoringSSL doesn't support multiple certificates replacing previously
  configured certificates on SSL_CTX_use_certificate(), this will match
  the current behaviour for disabled certificate compression
- AWS-LC just uses SSL_get_certificate(), supported since e1ba2b3e5 (v1.12.0)

pluknet commented:
compressBound -> deflateBound (a leftover from using compress(), NFC due to using default zlib settings)

pluknet commented:
For the record, I decided to not inline deflateBound() due to different behaviour in zlib vs zlib-ng, they once changed it in zlib-ng 70f608c7

pluknet committed:
Version bump.

pluknet committed:
Updated OpenSSL and PCRE used for win32 builds.

pluknet commented:
Added aux commits from #921 that don't actually need review (tired of waiting)
